R-package/R/lightgbm.R

```diff
@@ -114,12 +126,14 @@ NULL
 #'     }
 #' @inheritSection lgb_shared_params Early Stopping
 #' @return a trained \code{lgb.Booster}
+#' @importFrom parallel detectCores
 #' @export
 lightgbm <- function(data,
                      label = NULL,
                      weight = NULL,
                      params = list(),
                      nrounds = 100L,
+                     nthreads = parallel::detectCores(),
```

Comment:
Inserting this argument in the middle of the function signature is a breaking change, from the perspective of anyone calling `lightgbm::lightgbm()` using positional arguments.

Please move this to the bottom of the argument list (just before `...`) and move the corresponding parameter documentation in the roxygen comments to the same position in the function's documentation.

------

Per the conversation in #4192, https://github.com/microsoft/LightGBM/issues/4425#issuecomment-872943848, and elsewhere, LightGBM tends to run fastest when `num_threads` is set to the number of *physical* (not *logical*) CPUs. For `lightgbm::lightgbm()`, I recommend changing this to `parallel::detectCores(logical = FALSE)`.

---

Since this is being introduced as a top-level argument in the R package for the first time, please use the name `num_threads` instead of `nthreads`. I think it's desirable to have these keyword arguments match the "main" LightGBM parameters: https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_threads

Existing cases where that isn't true (e.g. `nrounds` instead of `num_iterations`) have just been preserved for backwards compatibility.

Reply:
Ok, will change it to `parallel::detectCores(logical = FALSE)`, however:
* In my own setup, I get better timings when fitting models with the full number of threads, despite the recommendation.
* The default for Python uses the full number of threads.
* Passing `logical=FALSE` doesn't always do what you're thinking, and it doesn't work in all operating systems.

Reply:
> In my own setup, I get better timings when fitting models with the full number of threads, despite the recommendation.

That's different from what we've observed in the past (e.g. https://github.com/microsoft/LightGBM/issues/4192#issuecomment-822023560, https://github.com/microsoft/LightGBM/issues/4425#issuecomment-872661139), but I guess it's dependent on the type of hardware you have, your operating system, version of relevant libraries, and where LightGBM is spending time in a particular training run.

> doesn't always do what you're thinking, and doesn't work in all operating systems

Oh interesting! I went researching for what you meant by this statement, and see the following in `?parallel::detectCores()`

```text
logical | if possible, use the number of physical CPUs/cores (if FALSE) or logical CPUs (if TRUE).
Currently this is honoured only on macOS, Solaris and Windows.
```

II also noticed this:

```text
An integer, NA if the answer is unknown.
```

Now that I see that it's possible for this function to return `NA`, I think that for maximum portability the default value should be set to `0L` (meaning the value of `omp_get_num_threads()`), instead of using any of R's mechanisms for detecting the available CPUs. That way, we wouldn't need to worry about an `NA` being passed through to that parameter (I'm not sure how that would be handled on the C++ side).

Setting the default to `0L` would also be consistent with the default on the C++ side:

* https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_threads
* https://github.com/microsoft/LightGBM/blob/60e72d5f4ec5ec2a4d30a53633e1e2d59c81309b/include/LightGBM/config.h#L205

And technically consistent with the default in the `scikit-learn` estimators in the Python package, where a value of `-1` is passed, since any value `<= 0` is treated by LightGBM as "take the result of `omp_get_num_threads()`".

* https://github.com/microsoft/LightGBM/blob/60e72d5f4ec5ec2a4d30a53633e1e2d59c81309b/python-package/lightgbm/sklearn.py#L365
* https://github.com/microsoft/LightGBM/blob/60e72d5f4ec5ec2a4d30a53633e1e2d59c81309b/include/LightGBM/utils/openmp_wrapper.h#L30-L34
* 

Please just change the default to `0L` and remove `{parallel}` from this PR. Thanks for pointing me towards the fact that `detectCores(logical = FALSE)` has different behavior on different systems!

Reply:
Can you tell us *why* you think that function (which would mean adding a new dependency to `{lightgbm}`) is preferable to setting this to `0L` and deferring to LightGBM's C++ code to pick a default?

Reply:
Because it ends up picking the actual number of physical cores even in setups in which `detectCores` would get it incorrectly, and never returns `NA`. I guess that the most typical scenario is to want to parallelize as much as possible, and in such case it'd be better to have the number of physical cores as the default and auto-detected correctly.

Reply:
Got it, thanks for that. I'm nervous about taking on a new dependency just for the purpose of setting this default, but I do think it's valuable to have `lightgbm::lightgbm()` by default choose the same `num_threads` as you'd find recommended in the documentation or issues page.

And since `{RhpcBLASctl}` is on CRAN (https://cran.r-project.org/web/packages/RhpcBLASctl/index.html), we at least know it meets CRAN's rigorous standards for portability.

I'm willing to try this. To avoid imposing a new *required* dependency on users of the R package, I'm suggesting the following:

* make `num_threads = NULL` the default in `lightgbm::lightgbm()`
* add a function in https://github.com/microsoft/LightGBM/blob/master/R-package/R/utils.R like the following:
    - ```r
       .suggest_default_num_threads <- function(){
           if (require("RhcpBLASctl")) {
              return(RhcpBLASctl::get_num_cores())
          } else {
              warning("Optional package RhcpBLASctl not found. Consider
          }
          return(0L)
       }
       ```
* in `lightgbm::lightgbm()`, have code like the following
    - ```r
       if (is.null(num_threads)) {
           num_threads <- .suggest_default_num_threads()
       }
       params <- lgb.check.wrapper_param(
          main_param_name = "num_threads"
          , params = params
          , alternative_kwarg_value = num_threads
        )
       ```
* add `{RhcpBLASctl}` as a `Suggests`-level import in https://github.com/microsoft/LightGBM/blob/master/R-package/DESCRIPTION

This would avoid adding a new required dependency, and would mean that the behavior of passing `num_threads = -1` or `num_threads = 0` through keyword arg `num_threads` would have the exact same meaning, the one documented inhttps://lightgbm.readthedocs.io/en/latest/Parameters.html#num_threads and described in my comments above in this thread.

That would also encapsulate this logic so that it could optionally be re-used in the future in other parts of the package (like in setting default `num_threads` for `Dataset` construction).

Reply:
But that would still require you to put that package under `Suggests`, and if something were to go wrong with that package on CRAN, you'd get the same kinds of issues as if it were a dependency.

Having the value wrapped like that is IMO less transparent to the user and less intuitive to see in a function signature.

Reply:
> and if something were to go wrong with that package on CRAN, you'd get the same kinds of issues as if it were a dependency

Sure, *on CRAN* a `Suggests` dependency imposes the same risks as an `Imports` or `Depends` one since CRAN will install all `Suggests` dependencies when running `R CMD check`.

But putting that into `Suggests` means that for users of the package, `install.packages("lightgbm")` won't require the installation of `{RhcpBLASctl}`, so `{RhcpBLASctl}` being incompatible with their environments wouldn't prevent the use of `{lightgbm}`.

I'm willing to take on a bit more maintenance burden (increased risk of CRAN failures caused by a dependency) in exchange for a better user experience with the package's default values, but I'm not willing to add a new unconditional dependency to `{lightgbm}` just to set a better default value for `num_threads`.